# Lab1 - Scikit-learn
Author: Christopher DiMattia

## 1. Introduction

The goal of this lab is to become familiar with the scikit-learn library.

You will practice loading example datasets, perform classification and regression with linear scikit-learn models, and investigate the effects of reducing the number of features (columns in X) and the number of samples (rows in X and y).


In [18]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

## 2. Classification

Using yellowbrick spam - classification  
https://www.scikit-yb.org/en/latest/api/datasets/spam.html

The goal is to investigate `LogisticRegression(max_iter=2000)` and effects of reducing the number of features and number of samples on classification performance.

### 2.1 Implement convenience function

In [19]:
from sklearn.metrics import accuracy_score
from sklearn.model_selection import train_test_split

def get_classifier_accuracy(model, X, y):
    '''Calculate train and validation accuracy of classifier (model)
        
        Splits feature matrix X and target vector y 
        with sklearn train_test_split() and random_state=956.
        
        model (sklearn classifier): Classifier to train and evaluate
        X (numpy.array or pandas.DataFrame): Feature matrix
        y (numpy.array or pandas.Series): Target vector
        
        returns: training accuracy, validation accuracy
    
    '''

    #split data into training/validaiton
    X_train, X_val, y_train, y_val = train_test_split(X, y, random_state=956)

    #fit model
    model.fit(X_train,y_train)

    #predict on training dataa
    y_train_predict = model.predict(X_train)
    #compare predicted to training data to get training accuracy
    accuracy_train = accuracy_score(y_train, y_train_predict)

    #predict on validation data
    y_val_predict = model.predict(X_val)
    #compare predicted to validation data to get validation accuracy
    accuracy_val = accuracy_score(y_val,y_val_predict)

    #return training and validation accuracies
    return(accuracy_train, accuracy_val)

### 2.2 Load data

Use the yellowbrick function `load_spam()`, load the spam data set into feature matrix `X` and target vector `y`.

Print size and type of `X` and `y`.


In [20]:
from yellowbrick.datasets.loaders import load_spam

#I'm not sure if you wanted the actual size function or the dimensions so I put both.  I also don't know what is meant by type (I assume data type).
X, y = load_spam()
print("X size: " + str(X.size) + ".  X types: float & int")
print("y size: " + str(y.size) + ".  Y type is boolean (spam or not spam represented as 0 or 1)")
print("X dimensions: " + str(X.shape))
print("y dimensions: " + str(y.shape))
print("X types " + str(X.dtypes))
print("Y type " + str(y.dtypes))

X size: 262200.  X types: float & int
y size: 4600.  Y type is boolean (spam or not spam represented as 0 or 1)
X dimensions: (4600, 57)
y dimensions: (4600,)
X types word_freq_make                float64
word_freq_address             float64
word_freq_all                 float64
word_freq_3d                  float64
word_freq_our                 float64
word_freq_over                float64
word_freq_remove              float64
word_freq_internet            float64
word_freq_order               float64
word_freq_mail                float64
word_freq_receive             float64
word_freq_will                float64
word_freq_people              float64
word_freq_report              float64
word_freq_addresses           float64
word_freq_free                float64
word_freq_business            float64
word_freq_email               float64
word_freq_you                 float64
word_freq_credit              float64
word_freq_your                float64
word_freq_font                float

Using the sklearn function `train_test_split()` prepare a feature matrix `X_small` and target vector `y_small` that contain only **1%** of the rows. Use `random_state=174`.

Print size and type of `X_small` and `y_small`.

In [21]:
X_small, X_val, y_small, y_val = train_test_split(X, y, random_state=174, train_size=0.01)
print("X_small size: " + str(X_small.size) + ".  X types: float & int")
print("y_small size: " + str(y_small.size) + ".  Y type is boolean (spam or not spam represented as 0 or 1)")
print("X_small dimensions: " + str(X_small.shape))
print("y_small dimensions: " + str(y_small.shape))
print("X_small types " + str(X_small.dtypes))
print("Y_small type " + str(y_small.dtypes))

X_small size: 2622.  X types: float & int
y_small size: 46.  Y type is boolean (spam or not spam represented as 0 or 1)
X_small dimensions: (46, 57)
y_small dimensions: (46,)
X_small types word_freq_make                float64
word_freq_address             float64
word_freq_all                 float64
word_freq_3d                  float64
word_freq_our                 float64
word_freq_over                float64
word_freq_remove              float64
word_freq_internet            float64
word_freq_order               float64
word_freq_mail                float64
word_freq_receive             float64
word_freq_will                float64
word_freq_people              float64
word_freq_report              float64
word_freq_addresses           float64
word_freq_free                float64
word_freq_business            float64
word_freq_email               float64
word_freq_you                 float64
word_freq_credit              float64
word_freq_your                float64
word_freq_fon

### 2.3 Train and evaluate models

1. Import `LogisticRegression` from sklearn
2. Instantiate model `LogisticRegression(max_iter=2000)`.
3. Create a pandas DataFrame `results` with columns: Data size, training accuracy, validation accuracy
4. Call your convenience function `get_classifier_accuracy()` using 
    - `X` and `y`
    - Only first two columns of `X` and `y`
    - `X_small` and `y_small`
5. Add the data size, training and validation accuracy for each call to the `results` DataFrame
6. Print `results`

In [23]:
from sklearn.linear_model import LogisticRegression

#create model
logRegression = LogisticRegression(max_iter=2000)

#get accuracies useing the convenience function
X_y = get_classifier_accuracy(logRegression,X,y)
X_y_2col = get_classifier_accuracy(logRegression,X.iloc[:,0:2],y)
X_y_small = get_classifier_accuracy(logRegression,X_small,y_small)

#add labels for each datasize and include difference in accuracy between validation and training to help answer next question
X_y = ("X & y"),X_y[0],X_y[1],X_y[0]-X_y[1]
X_y_2col = ("X & y w/ 2 features"),X_y_2col[0],X_y_2col[1],X_y_2col[0]-X_y_2col[1]
X_y_small = ("X & y, small"),X_y_small[0],X_y_small[1],X_y_small[0]-X_y_small[1]

#list of tuples to input into dataframe
data = [X_y,X_y_2col,X_y_small]

#create dataframe and fill with results
results = pd.DataFrame(data, columns=["DataSize","Training Accuracy","Validation Accuracy","TRain Acc minus Val Acc"])
#print/show results
results

DataSize  Training Accuracy  Validation Accuracy  \
0                X & y           0.935072             0.917391   
1  X & y w/ 2 features           0.608986             0.613043   
2         X & y, small           0.941176             0.750000   

   TRain Acc minus Val Acc  
0                 0.017681  
1                -0.004058  
2                 0.191176

### 2.4 Questions
1. What is the validation accuracy using all data? What is the difference between training and validation accuracy?
2. How does the validation accuracy and difference between training and validation change when only two columns are used? Provide values.
3. How does the validation accuracy and difference between training and validation change when only 1% of the rows are used? Provide values.

Answer
1. Validation accuracy with all data: 91.7%.  Training accuracy is 1.8% higher
2. The validation accuracy drops drastically to 61.3& (a 30.4% drop) but the difference between the training and validation accuracy is still relatively small at 0%
3. The validation accuracy greatly drops from 91.7% to 75% but the difference between the training and validation accuracy drastically increases to a 19.1% difference
See above

## 3. Regression

Using yellowbrick energy - regression  
https://www.scikit-yb.org/en/latest/api/datasets/energy.html

The goal is to investigate `LinearRegression()` and effects of reducing the number of features and number of samples on regression performance.

### 3.1 Implement convenience function

In [24]:
from sklearn.metrics import mean_squared_error
from sklearn.model_selection import train_test_split

def get_regressor_mse(model, X, y):
    '''Calculate train and validation mean-squared error (mse) of regressor (model)
        
        Splits feature matrix X and target vector y 
        with sklearn train_test_split() and random_state=956.
        
        model (sklearn regressor): Regressor to train and evaluate
        X (numpy.array or pandas.DataFrame): Feature matrix
        y (numpy.array or pandas.Series): Target vector
        
        returns: training mse, validation mse
    
    '''
    #same as above
    #split data into training/validaiton
    X_train, X_val, y_train, y_val = train_test_split(X, y, random_state=956)

    #fit model
    model.fit(X_train,y_train)

    #predict on training dataa
    y_train_predict = model.predict(X_train)
    #same as classification except this time it's regression so using mse accuracy to get accuracy
    mse_train = mean_squared_error(y_train, y_train_predict)

    #predict on validation data
    y_val_predict = model.predict(X_val)
    #compare predicted to validation data to get mse
    mse_val = mean_squared_error(y_val,y_val_predict)

    #return training and validation accuracies
    return(mse_train, mse_val)
   

    

### 3.2 Load data

Use the yellowbrick function `load_energy()` load the energy data set into feature matrix `X` and target vector `y`.

Print dimensions and type of `X` and `y`.

In [25]:
from yellowbrick.datasets.loaders import load_energy

#Similar to above
X, y = load_energy()
print("X size: " + str(X.size) + ".  X types: float & int")
print("y size: " + str(y.size) + ".  Y type is boolean (spam or not spam represented as 0 or 1)")
print("X dimensions: " + str(X.shape))
print("y dimensions: " + str(y.shape))
print("X types " + str(X.dtypes))
print("Y type " + str(y.dtypes))

X size: 6144.  X types: float & int
y size: 768.  Y type is boolean (spam or not spam represented as 0 or 1)
X dimensions: (768, 8)
y dimensions: (768,)
X types relative compactness         float64
surface area                 float64
wall area                    float64
roof area                    float64
overall height               float64
orientation                    int64
glazing area                 float64
glazing area distribution      int64
dtype: object
Y type float64


Using the sklearn function `train_test_split()` prepare a feature matrix `X_small` and target vector `y_small` that contain only **1%** of the rows. Use `random_state=174`.

Print size and type of `X_small` and `y_small`.

In [26]:
#Same as above.  Split the data into training and validation sets
X_small, X_val, y_small, y_val = train_test_split(X, y, random_state=174, train_size=0.01)
print("X_small size: " + str(X_small.size) + ".  X types: float & int")
print("y_small size: " + str(y_small.size) + ".  Y type is boolean (spam or not spam represented as 0 or 1)")
print("X_small dimensions: " + str(X_small.shape))
print("y_small dimensions: " + str(y_small.shape))
print("X_small types " + str(X_small.dtypes))
print("Y_small type " + str(y_small.dtypes))

X_small size: 56.  X types: float & int
y_small size: 7.  Y type is boolean (spam or not spam represented as 0 or 1)
X_small dimensions: (7, 8)
y_small dimensions: (7,)
X_small types relative compactness         float64
surface area                 float64
wall area                    float64
roof area                    float64
overall height               float64
orientation                    int64
glazing area                 float64
glazing area distribution      int64
dtype: object
Y_small type float64


### 3.3 Train and evaluate models

1. Import `LinearRegression` from sklearn
2. Instantiate model `LinearRegression()`.
3. Create a pandas DataFrame `results` with columns: Data size, training MSE, validation MSE
4. Call your convenience function `get_regressor_mse()` using 
    - `X` and `y`
    - Only first two columns of `X` and `y`
    - `X_small` and `y_small`
5. Add the data size, training and validation MSE for each call to the `results` DataFrame
6. Print `results`

In [27]:
from sklearn.linear_model import LinearRegression

#create model
linearRegression = LinearRegression()

#get accuracies useing the convenience function
X_y = get_regressor_mse(linearRegression,X,y)
X_y_2col = get_regressor_mse(linearRegression,X.iloc[:,0:2],y)
X_y_small = get_regressor_mse(linearRegression,X_small,y_small)

X_y

#add labels for each datasize and include difference in accuracy between validation and training to help answer next question
X_y = ("X & y"),X_y[0],X_y[1],X_y[0]-X_y[1]
X_y_2col = ("X & y w/ 2 features"),X_y_2col[0],X_y_2col[1],X_y_2col[0]-X_y_2col[1]
X_y_small = ("X & y, small"),X_y_small[0],X_y_small[1],X_y_small[0]-X_y_small[1]

#list of tuples to input into dataframe
data = [X_y,X_y_2col,X_y_small]

#create dataframe and fill with results
results = pd.DataFrame(data, columns=["DataSize","Training MSE","Validation MSE","Train MSE minus Val MSE"])
#print/show results
results


DataSize  Training MSE  Validation MSE  Train MSE minus Val MSE
0                X & y  7.981975e+00       10.292306                -2.310331
1  X & y w/ 2 features  5.360043e+01       46.410426                 7.190004
2         X & y, small  2.284541e-28       69.977449               -69.977449

### 3.4 Questions
1. What is the validation MSE using all data? What is the difference between training and validation MSE?
1. How does the validation MSE and difference between training and validation change when only two columns are used? Provide values.
1. How does the validation MSE and difference between training and validation change when only 1% of the rows are used? Provide values.

*YOUR ANSWERS HERE*
1. The validation MSE for all data is 10.3.  The difference between the two is -2.3 (higher validation MSE)
2. When only two columns are used the validation MSE increases (gets worse) from 10.3 to 46.4 while the difference in validation MSE and training MSE becomes larger at 7.2 (training  minus validation).
3. The validation MSE greatly increases from 10.3 to 70.0 while the difference in validation and training MSE drastically increases to -70.0 (training minus validation).

See above

## 4. Observations/Interpretation

Describe any pattern you see in the results. Relate your findings to what we discussed during lectures. Include data to justify your findings.


*ADD YOUR FINDINGS HERE*

Patterns:
1. More features obviously results in high accuracies (compare 92% vs 67% for classification and 10.3 vs 46.4 mse for regression) because the model is able to generalize better because it can utalize all the information from the features.
2. Larger data sets also obviously result in higher accuracies (compare 92% vs 75% for classification and 10.3 vs 70.0 for regression) and the inverse is true.  With a smaller dataset the model doesn't have enough information to generalize properly.
3. Smaller data sets will still retain high training accuracy () because even with a small amount of data a curve can still be fit quite well (we can conceive of an extreme example of only a few data points that can be fit perfectly).  In comparision removing most/all of a datas features will result in a low training accuracy which makes sense given that two features could be highly uncorrelated or even antagonistic to each other.  With so few features the model is not able to fully generalize as it is missing key information from other features so validation accuracy is clearly going to be poor.  Unless those 2 specific features are overwhelmingly impactful the model is likely to have poor performance as most models attempt to generalize complex infomation and relationships.


## 5. Reflection
Include a sentence or two about:
- what you liked or disliked,
- found interesting, confusing, challangeing, motivating
while working on this assignment.


*ADD YOUR THOUGHTS HERE*
Likes
1. I greatly liked the ease of which I could create and get results from the models given how long the actual mathematical implmentation is.  I particularily like how the way sklearn is set up in that I am able to swap out any models with ease to compare different ML model fits.
2. I liked having a large dataset that was non-trivial.  While trivial datasets are excellent for learning it's a nice change of pace to use a larger amount of data.  It made me realize that interpretting results will be far more challenging than with simple assignments and while this assignment was not heavily focused on interpretation it does make it clear that interpreting results with many features will pose a challenge.

Dislikes
1. Not a dislike about the assignment in particular but due to how other assignments lined up this week I did not have a chance to test other models just to see for comparision.  I want to do more comparisions between different types of models to get a better handle of the advantages/disadvantages of each via direct comparision.
2. While I liked the ease with which the models are implemented I do think it would be beneficial to create a model from scratch so as to better appreciate the math behind the model.  I also think it would be interesting to do a "speed" comparision between these highly optimized models and a laymen's implmentation.
3. I didn't get a sense for the weights that were applied to each feature.  Had I had more time I would have liked to print those (I assume I can) and try to see if I could identify or at least get a grasp of what features most impact the results.
4. I'm not sure it would be useful but perhaps more plotting of results would be intesting for interpretation (to see how useful/useless it is).  That said I can do this on my own.

Interesting, Confusing, etc
1.  I found the assignment fairly straight forward with little confusion or challenge.  The most difficult part was the interpretation as that required the most critical thought and I'm still not sure I captured/know all the main theoretical points.  I for one struggle to conceptualize how models perform when the features/data sizes are manipulated beyong the obvious conclusions (Alot is obvious such as "more data is obviously better", but certain queestions I don't have a solid answer for.  An example: Can you create two different models that have similar performance and the exact same data but completely different weighting/interpretations?  I'm sure there are methods to address this but I haven't looked into them).  I also found it very interesting and motivating that the accuracy for the spam data was so high (91% with all data) given how complex natural langauge is, espcially given these are probably the simplest ML techniques.